### ISSUE TITLE
Support for pulling metadata from HTML index pages

### ISSUE THREAD

**Issue by kennethjor** [rclone/rclone#7357]
#### The associated forum post URL from `https://forum.rclone.org`
None

#### What is your current rclone version (output from `rclone version`)?
Slightly older version `v1.60.1-DEV`, but I've checked the source code for the [HTTP backend](https://github.com/rclone/rclone/tree/master/backend/http) and don't see anything in there.

#### What problem are you are trying to solve?
The beta archive (https://beta.rclone.org/v1.61.0-beta.6504.db5d58240/) for rclone itself contains HTML markup from which file sizes and timestamps could be extracted. However, rclone does not currently support this without a bunch of `HEAD` requests.
```
$ rclone lsl --http-url https://beta.rclone.org/ --http-no-head :http:v1.61.0-beta.6504.db5d58240/
       -1 1970-01-01 09:00:00.000000000 rclone-v1.61.0-beta.6504.db5d58240-freebsd-386.zip
       -1 1970-01-01 09:00:00.000000000 rclone-v1.61.0-beta.6504.db5d58240-freebsd-amd64.zip
       -1 1970-01-01 09:00:00.000000000 rclone-v1.61.0-beta.6504.db5d58240-freebsd-arm-v7.zip
       -1 1970-01-01 09:00:00.000000000 rclone-v1.61.0-beta.6504.db5d58240-freebsd-arm.zip
<cut>
```

#### How do you think rclone should be changed to solve that?
Would it be possible to come up with some simple expecation for HTML when, if met, allows rclone to extract metadata from the markup if `--http-no-head` is set?

For instance, taking inspiration from `beta.rclone.org`, it has the following (simplified) table row:
```
<tr>
	<td><a href="./rclone-v1.61.0-beta.6504.db5d58240-freebsd-386.zip">rclone-v1.61.0-beta.6504.db5d58240-freebsd-386.zip</a></td>
	<td>15 MiB</td>
	<td><time datetime="2022-10-21T15:35:20Z">22/10/2022, 00:35:20</time></td>
</tr>
```
rclone could assume that, if the anchor is contained within a table row, the other columns contain information about the file. The `<time/>` tag is easily parsed, but it's not clear how file size should be defined.

Alternatively, a much simpler solution would be to rely on the anchor tag only and store metadata in `data-*` attributes, like:
```
<a href="./rclone-v1.61.0-beta.6504.db5d58240-freebsd-386.zip" data-datetime="2022-10-21T15:35:20Z" data-size="15932609">rclone-v1.61.0-beta.6504.db5d58240-freebsd-386.zip</a>
```
Although, this would require index pages to explicitly add support for it, but wouldn't require rclone to traverse the DOM as in the first example.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
This would need a parser for every type of file listing in html. You can see the files we have here that we use for testing

https://github.com/rclone/rclone/tree/master/backend/http/test/index_files

So not impossible but a lot of work!

I note that the listings don't always have the full metadata - they might have a rounded size, eg `1.0M` which isn't going to work well with rclone.

**Comment by kennethjor** [comment]
@ncw Yeah, I did study those example files in more detail after I submitted the ticket. I agree it would not be feasible parse all of these reliably.

I'd like to propose the second option in my original post, echoed here for clarity:
```
<a href="./rclone-v1.61.0-beta.6504.db5d58240-freebsd-386.zip"
    data-file-time="2022-10-21T15:35:20Z"
    data-file-size="15932609"
    >rclone-v1.61.0-beta.6504.db5d58240-freebsd-386.zip</a>
```
I did have a look, but was not able to find any real prior implementations, so this is not a standard I've pulled from anywhere. I'm suggesting we make this the "rclone way" and it's only activated when `--http-no-head` is set. This would allow anyone with their own index pages, wishing to have good and fast support for `rclone`, to add it if they wished.

The following constraints could be set for a simpler implementation:
* `data-file-time` MUST be a proper ISO-8601 timestamp, similar to what the `<time>` element allows, just more restrictive.
* `data-file-size` MUST be an positive integer with no non-numeric characters in it.

If either of these fields are present and are valid, and `--http-no-head` is used, they will be used for the indexing in rclone. If not provided or containing invalid values, they will simply be ignored and revert to defaults.

My personal use case is I have a [data site](https://data.everef.net/) with a lot of files available for download. I build my own index files. I think my users would find value in being able to mount it with `rclone` and have fast access to the indexes, without the need for doing a bunch of `HEAD` requests.

**Comment by ncw** [comment]
I guess we could make this generic using something like css selector syntax (nice library: https://github.com/ericchiang/css ) to find the relevant node in the HTML tree (which rclone has parsed already) and we could tell rclone we want to read the attribute value or the contents of the node. Rclone would need a hint about time format.

So taking caddy's output from beta.rclone.org

```
<tr class="file">
	<td></td>
	<td>
		<a href="./version.txt">
			<svg width="1.5em" height="1em" version="1.1" viewBox="0 0 265 323"><use xlink:href="#file"></use></svg>
			<span class="name">version.txt</span>
		</a>
	</td>
	<td data-order="35">35 B</td>
	<td class="hideable"><time datetime="2023-10-10T15:15:46Z">10/10/2023 03:15:46 PM +00:00</time></td>
	<td class="hideable"></td>
</tr>
```

We could specify using CSS selectors the URL then use relative CSS selectors to find the other data. That wold give a simple bit of config which could be used to parse any site.
